### PR TITLE
#1966 - Highlight games with friends

### LINKED ISSUE #1966: Indicate that a game include friends of your
**Is your feature request related to a problem? Please describe.**
You are on Discord with your friends and they say they are in a game and it takes you 2 minutes till you understood which game you are supposed to join.

**Describe the solution you'd like**
Highlight the games a friend is in. Possibly with a smiley, the same that is used to indicate somebody is a friend in the player_card_tooltip.fxml (style class friend-icon). Though getting a better icon would also be appreciated. Needs to be done both in the game tiles and game table.

**Wanna have the bug fixed quickly?**
Visit Issue hunt...
[Issue hunt](https://issuehunt.io/r/FAForever/downlords-faf-client)

### PR DESCRIPTION
**Pull request by Marc-Spector**
Closes #1966

### CODE REVIEW

**Comment by Marc-Spector** [comment]
<details>
  <summary>Demo</summary>

demo


</details>

**Comment by Sheikah45** [comment]
Personally I am not really a fan of adding another check box to the top of the games table especially as it has a rather long name.

**Comment by Marc-Spector** [comment]
@Sheikah45 Any ideas or suggestions?

**Comment by Sheikah45** [comment]
I wonder if just having it always been on as like a property of the game would be good or adding an icon that can just be displayed like the lock for private games

**Comment by Marc-Spector** [comment]
@Sheikah45 Maybe simply to paint current/max players label to green color?

**Comment by Marc-Spector** [comment]
I do not want to add the icons because the table is overloaded

**Comment by Sheikah45** [comment]
Maybe or the title. But I don't mind the background highlighting you already have. Although it may not be apparent what it means when people first see it

File: src/main/resources/theme/style.css [hunk 1]
```diff
@@ -1177,11 +1177,23 @@
   -fx-font-size: 1.2em;
 }
 
+.game-card:friendInGame {
+  -fx-background-color: rgba(42, 253, 53, 0.15);
```
Comment (1-alex98) [inline src/main/resources/theme/style.css]: Define colour in colours.css

**Comment by Marc-Spector** [comment]
> Maybe or the title. But I don't mind the background highlighting you already have. Although it may not be apparent what it means when people first see it

The title and number of players with green color looked unethical. So I decided to leave the background highlighting unchanged.

File: src/main/java/com/faforever/client/player/PlayerService.java [hunk 2]
```diff
@@ -214,6 +202,19 @@ private void updateGameDataForPlayer(Game game, Player player) {
     updatePlayerChatUsers(player);
   }
 
+  public boolean areFriendsInGame(Game game) {
+    if (game == null) {
+      return false;
+    }
+    Map<String, List<String>> teams = game.getTeams();
+    synchronized (game.getTeams()) {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerService.java]: I would synchronize then get the teams

File: src/main/java/com/faforever/client/player/PlayerService.java [hunk 3]
```diff
@@ -101,23 +91,21 @@ public void onGameRemoved(GameRemovedEvent event) {
...
     synchronized (game.getTeams()) {
-      List<String> playersInGame = teams.entrySet().stream()
-          .flatMap(stringListEntry -> stringListEntry.getValue().stream())
-          .collect(Collectors.toList());
-      updateGamePlayers(playersInGame, game);
+      updateGamePlayers(getPlayersNamesOfAllTeams(teams), game);
     }
   }
 
+  private List<String> getPlayersNamesOfAllTeams(Map<String, List<String>> teams) {
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/player/PlayerService.java]: I would rename to getAllPlayerNames

**Review by Sheikah45 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/game/GamesTableController.java [hunk 4]
```diff
@@ -209,8 +214,11 @@ protected void updateItem(Game game, boolean empty) {
         super.updateItem(game, empty);
         if (empty || game == null) {
           setTooltip(null);
+          pseudoClassStateChanged(FRIEND_IN_GAME_PSEUDO_CLASS, false);
         } else {
           setTooltip(tooltip);
+          pseudoClassStateChanged(FRIEND_IN_GAME_PSEUDO_CLASS, playerService.areFriendsInGame(game)
+                  && !coopMissionName.isVisible()); // do not highlight games in coop tab
```
Comment (Sheikah45) [inline src/main/java/com/faforever/client/game/GamesTableController.java]:
```suggestion
                  && game.getGameType() != GameType.COOP
```

**Review by Sheikah45 — CHANGES_REQUESTED** [review]
Small change on how coop games are detected

**Comment by Sheikah45** [comment]
Went ahead and changed it myself

**Review by Sheikah45 — APPROVED** [review]

**Pull request merged**